### PR TITLE
Feature/fuerte cluster 9000

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

Make more code use fuerte networking

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as all of the existing unit tests

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/8332/

### CODE REVIEW

File: arangod/RestHandler/RestEdgesHandler.cpp [hunk 1]
```diff
@@ -141,17 +91,46 @@ bool RestEdgesHandler::parseDirection(TRI_edge_direction_e& direction) {
...
+  auto bindParameters = std::make_shared<VPackBuilder>();
+  bindParameters->openObject();
+  bindParameters->add("@collection", VPackValue(cname));
+  bindParameters->add("vertex", VPackValue(vertexId));
+  bindParameters->close();
+  auto options = std::make_shared<VPackBuilder>();
+
+  arangodb::aql::Query query(false, vocbase, aql::QueryString(queryString(dir)),
+                             bindParameters, options, arangodb::aql::PART_MAIN);
+  query.queryOptions().cache = false;
```
Comment (jsteemann) [inline arangod/RestHandler/RestEdgesHandler.cpp]: Why is the query cache turned off here? I don't think this is necessary.

File: arangod/RestHandler/RestEdgesHandler.cpp [hunk 2]
```diff
@@ -141,17 +91,46 @@ bool RestEdgesHandler::parseDirection(TRI_edge_direction_e& direction) {
...
+aql::QueryResult queryEdges(TRI_vocbase_t& vocbase, std::string const& cname,
+                            TRI_edge_direction_e dir, std::string const& vertexId) {
+  auto bindParameters = std::make_shared<VPackBuilder>();
+  bindParameters->openObject();
+  bindParameters->add("@collection", VPackValue(cname));
+  bindParameters->add("vertex", VPackValue(vertexId));
+  bindParameters->close();
+  auto options = std::make_shared<VPackBuilder>();
+
+  arangodb::aql::Query query(false, vocbase, aql::QueryString(queryString(dir)),
```
Comment (jsteemann) [inline arangod/RestHandler/RestEdgesHandler.cpp]: This method is not using `RestVocBaseHandler::createTransaction()` anymore as the previous version did.
Wondering if we can still use the edges functionality from inside a cluster transaction.
IIRC then aql::Query can't pick up an existing cluster transaction, so these operations now are run outside of existing cluster transactions, even if the transaction id header is passed to the RestHandler.

I think we should also add tests to verify the transaction behavior and ensure it doesn't break over time.

File: arangod/RestHandler/RestEdgesHandler.cpp [hunk 3]
```diff
@@ -327,103 +247,52 @@ bool RestEdgesHandler::readEdgesForMultipleVertices() {
...
-      // If we are a DBserver, we want to use the cluster-wide collection
-      // name for error reporting:
-      collectionName = trx->resolver()->getCollectionNameCluster(cid);
+    VPackSlice edges = queryResult.data->slice();
+    for (VPackSlice edge : VPackArrayIterator(edges)) {
+      std::string key = transaction::helpers::extractKeyFromDocument(edge).copyString();
+      if (foundEdges.find(key) == foundEdges.end()) {
+        resultBuilder.add(edge);
+        foundEdges.emplace(std::move(key));
+      }
```
Comment (jsteemann) [inline arangod/RestHandler/RestEdgesHandler.cpp]: If we don't expect too many duplicates, we could simplify this to just
```
      if (foundEdges.emplace(transaction::helpers::extractKeyFromDocument(edge).copyString()).second) {
        resultBuilder.add(edge);
      }
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 4]
```diff
@@ -1189,80 +1162,79 @@ int selectivityEstimatesOnCoordinator(ClusterFeature& feature, std::string const
...
+          }
+        }
 
-  auto aggregate_indexes = [](std::vector<double> vec) -> double {
-    TRI_ASSERT(!vec.empty());
-    double rv = std::accumulate(vec.begin(), vec.end(), 0.0);
-    rv /= static_cast<double>(vec.size());
-    return rv;
-  };
+        auto aggregate_indexes = [](std::vector<double> vec) -> double {
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp]:
```suggestion
        auto aggregate_indexes = [](std::vector<double> const& vec) -> double {
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 5]
```diff
@@ -1920,90 +1977,91 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+          auto buffer = r.response->stealPayload();
+
+          VPackSlice resSlice(buffer->data());
+          if (!resSlice.isObject()) {
+            // Response has invalid format
+            THROW_ARANGO_EXCEPTION(TRI_ERROR_HTTP_CORRUPTED_JSON);
+          }
+          if (r.response->statusCode() != fuerte::StatusOK) {
+            int code = arangodb::basics::VelocyPackHelper::getNumericValue<int>(
+                resSlice, "errorNum", TRI_ERROR_INTERNAL);
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp]:
```suggestion
                resSlice, StaticStrings::ErrorNum, TRI_ERROR_INTERNAL);
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 6]
```diff
@@ -1920,90 +1977,91 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
 
-  std::vector<ClusterCommRequest> requests;
-  auto body = std::make_shared<std::string>(builder.toJson());
+  auto* pool = trx.vocbase().server().getFeature<NetworkFeature>().pool();
+
+  std::vector<Future<network::Response>> futures;
+  futures.reserve(engines->size());
+
+  VPackBufferUInt8 buffer;
+  buffer.append(lease->data(), lease->size());
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp]: Why is it necessary to copy everything from `lease` into `buffer` here?
`buffer` is copied when we call `network::sendRequestRetry` and not modified, so we could instead pass the already existing velocypack::Buffer of `lease` into `network::sendRequestTry`, and get rid of `buffer` entirely.

File: arangod/Cluster/ClusterMethods.cpp [hunk 7]
```diff
@@ -2050,191 +2104,64 @@ void fetchVerticesFromEngines(
...
-  return TRI_ERROR_NO_ERROR;
+          auto payload = r.response->stealPayload();
+          VPackSlice resSlice = VPackSlice(payload->data());
+          if (!resSlice.isObject()) {
+            // Response has invalid format
+            THROW_ARANGO_EXCEPTION(TRI_ERROR_HTTP_CORRUPTED_JSON);
+          }
+          if (r.response->statusCode() != fuerte::StatusOK) {
+            int code = arangodb::basics::VelocyPackHelper::getNumericValue<int>(
+                resSlice, "errorNum", TRI_ERROR_INTERNAL);
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp]:
```suggestion
                resSlice, StaticStrings::ErrorNum, TRI_ERROR_INTERNAL);
```

File: arangod/RestHandler/RestEdgesHandler.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/RestHandler/RestEdgesHandler.cpp]: this handler never supported el-cheapo. if the docs state it otherwise the docs are lying

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/8236/

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/8332/

File: arangod/RestHandler/RestEdgesHandler.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/RestHandler/RestEdgesHandler.cpp]: No, it wasn't documented. But is there a problem with supporting transactions here?

Comment (graetzer) [reply arangod/RestHandler/RestEdgesHandler.cpp]: yes

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/8342/

File: arangod/RestHandler/RestEdgesHandler.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/RestHandler/RestEdgesHandler.cpp]: no need to improve APIs that should be deprecated anyway

**Comment by jsteemann** [comment]
Adding a full run for extra logID tests:
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6642/

**Pull request merged**